include/zenoh-pico/collections/refcount.h

```diff
@@ -0,0 +1,261 @@
+//
+// Copyright (c) 2022 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+#ifndef ZENOH_PICO_COLLECTIONS_REFCOUNT_H
+#define ZENOH_PICO_COLLECTIONS_REFCOUNT_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#if Z_FEATURE_MULTI_THREAD == 1
+#if ZENOH_C_STANDARD != 99
+
+#ifndef __cplusplus
+#include <stdatomic.h>
+#define _z_atomic(X) _Atomic X
+#define _z_atomic_store_explicit atomic_store_explicit
+#define _z_atomic_fetch_add_explicit atomic_fetch_add_explicit
+#define _z_atomic_fetch_sub_explicit atomic_fetch_sub_explicit
+#define _z_memory_order_acquire memory_order_acquire
+#define _z_memory_order_release memory_order_release
+#define _z_memory_order_relaxed memory_order_relaxed
+#else
+#include <atomic>
+#define _z_atomic(X) std::atomic<X>
+#define _z_atomic_store_explicit std::atomic_store_explicit
+#define _z_atomic_fetch_add_explicit std::atomic_fetch_add_explicit
+#define _z_atomic_fetch_sub_explicit std::atomic_fetch_sub_explicit
+#define _z_memory_order_acquire std::memory_order_acquire
+#define _z_memory_order_release std::memory_order_release
+#define _z_memory_order_relaxed std::memory_order_relaxed
+#endif  // __cplusplus
+
+/*------------------ Internal Array Macros ------------------*/
+#define _Z_REFCOUNT_DEFINE(name, type)                                                          \
+    typedef struct name##_rc_t {                                                                \
+        type##_t *ptr;                                                                          \
```

Comment:
This should be a single pointer to `struct name##_rc_inner_t {type##_t value; _z_atomic(unsigned int) _cnt;};`, making the type smaller, reducing the number of allocations and reducing the number of cache lines occupied by a single value.